**Registration Number:** 230154977

# Part 1: Translation functions

###### Python version 3.12.1

#### Xeno-nomenclature

- **K-pop**: The alien language.
- **Boop**: A pair $(a, b)$ representing the fraction $\frac{a}{b}$.
- **Salient**: An operation between two boops defined by 
$$
(a, b) \boxplus (c, d) = (a + c, b + d)
$$
- **Spork**: A binary tree of boops generated by taking repeated salients.
- **Frond**: A single row of the spork.
- **n-Spork**: The first $n+1$ fronds of the spork.
- **Kebab**: A string of letters describing a path taken through the spork. Starting at the root of the spork, a letter 'E' (**Eep**) describes a right movement and a letter 'W' (**Woop**) describes a left movement. For example, the kebab 'WEEE' is one left movement and three right movements, which represents the boop $(4, 5)$. 
- **Pronk**: A pronk is an encoding of a kebab: a list of integers corresponding to the number of successive Eeps and Woops. The first number in the pronk is always the number of starting Eeps. For example, the kebab 'WEEE' can be written as the pronk $[0, 1, 3]$. 
- **Wumpus**: A wumpus is a pronk of infinite length. They can be used to represent irrational numbers.
- **Frood**: A frood is a truncation of a wumpus (so a finite pronk) which yields a good approximation of the irrational number.
- **Hoopy Frood**: A 'very good' approximation of an irrational number. They are obtained by truncating immediately before a number greater than $100$ in a wumpus.
- **Friendly [wumpus]**: A wumpus is 'friendly' if it eventually settles into a repeating pattern.
- **Hostile [wumpus]**: If a wumpus does not eventually reach a repeating pattern, it is called 'hostile'.
- **Mordant [wumpus]**: A wumpus is 'mordant' if it is bounded above by a finite integer.
- **Rampant [wumpus]**: If a wumpus has divergent terms, it is 'rampant'.

In [3]:
def boop_to_kebab(B: tuple[int, int]):
    """
    Convert a boop into a kebab. We expect large x/y to have 
    lots of eeps and a very small x/y to start with lots of woops.

    :param tuple B: Boop (x, y) representing the fraction x/y
    :returns: A string of Eeps (E) and Woops (W)
    :rtype: string 
    """
    x = B[0]
    y = B[1]

    S_list = []

    while x != y:
        if x > y:
            x = x - y
            S_list.append("E")
        else:
            y = y - x
            S_list.append("W")

    return "".join(S_list)


# Tests
# print(boop_to_kebab((213, 353)))

In [4]:
def kebab_to_boop(S: str):
    """
    Convert a kebab into a boop.

    :param string S: A kebab S, consisting of a series of W's and E's
    :returns: A boop
    :rtype: tuple[int, int]
    :raises ValueError: Invalid kebab
    """

    x, y = 1, 1
    kebab_list = list(S)

    while len(kebab_list) > 0:
        l = kebab_list.pop()

        if l == "E":
            x = x + y
        elif l == "W":
            y = y + x
        else:
            raise 
        ValueError("A kebab must only contain 'W' and 'S")

    return (x, y)


# Tests
# print(kebab_to_boop("EEW"))  # Output: (5, 2)
# print(kebab_to_boop(""))     # Output: (1, 1)
# print(kebab_to_boop("EEEWWWWWWWEEEEEEEEEEEEEEEW"))  # Output: (688, 219)
# print(kebab_to_boop("EEEWWWWWWWEEEEEEEEEEEEEEE"))   # Output: (355, 113)

In [5]:
def kebab_to_pronk(S: str):
    """
    Convert a kebab into pronk.

    :param string S: A kebab as a string. E.g. "WEWWEE"
    :returns: A pronk. An encoding of a kebab, with the first
    entry in the list being the number of eeps starting the kebab
    :rtype: list[int]
    :raises ValueError: Invalid kebab
    """

    curr = "E"
    next = "W"
    r_kebab = list( reversed(S))
    pronk = []

    c = 0
    while len(r_kebab) > 0:
        l = r_kebab.pop()

        if l == curr:
            c += 1
        elif l == next:
            curr, next = next, curr
            pronk.append(c)
            c = 1  # Reset to 1 as we have parsed this letter
        else:
             raise 
        ValueError("A kebab must only contain 'W' and 'S")

    # Append final sequence
    pronk.append(c)

    return pronk


# Tests
# print(kebab_to_pronk("WEEWEWEEWWE"))  # Output: [0, 1, 2, 1, 1, 1, 2, 2, 1]
# print(kebab_to_pronk("WWWWWW"))       # Output: [0, 6]

In [6]:
def pronk_to_kebab(L: list[int]):
    """
    Convert a pronk to a kebab.

    :param list[int] L: A pronk
    :returns: A kebab
    :rtype: string
    """

    return "".join([n * ("W" if i % 2 == 1 else "E") for i, n in enumerate(L)])


# Tests
# print(pronk_to_kebab([0, 3, 4, 2, 3]))  # Output: WWWEEEEWWEEE

# Part 2: Cracking the code

In [7]:
# First, let's decode the message!
boop1 = (137482210, 27112921)
boop2 = (65383333292, 8124730839)

# Helpers
pronk_to_str = lambda p: "".join([chr(n + 96) for n in p])
str_to_pronk = lambda s: [ord(c)-96 for c in list(s)]

def boop_to_english(B: tuple[int, int]):
    """
    Converts a boop to the word it represents in english

    :param tuple[int, int] B: A boop
    :returns: A string, which may be a valid english word
    :rtype: string
    """
    kebab = boop_to_kebab(B)
    pronk = kebab_to_pronk(kebab)
    english = pronk_to_str(pronk)
    return english


print(boop_to_english(boop1))
print(boop_to_english(boop2))


# The aliens have an admiration for Engelbert Humperdinck!
# His birth name is Arnold George Dorsey, so that's what we 
# need to transmit back to the aliens.


def english_to_boop(E: str):
    """
    Converts a string into a representative boop

    :param string E: A word
    :returns: A boop representing that word
    :rtype: tuple[int, int]
    """
    pronk = str_to_pronk(E)
    kebab = pronk_to_kebab(pronk)
    boop = kebab_to_boop(kebab)
    return boop


def transmit():
    """
    Transmit the words "arnold", "george" and "dorsey"
    to the aliens.
    """

    arnold_boop = english_to_boop("arnold")
    george_boop = english_to_boop("george")
    dorsey_boop = english_to_boop("dorsey")

    return [arnold_boop, george_boop, dorsey_boop]


response_boops = transmit()


# Tests
# print(response_boops)
# print(boop_to_english(response_boops[0]))  # Output: arnold
# print(boop_to_english(response_boops[1]))  # Output: george
# print(boop_to_english(response_boops[2]))  # Output: dorsey


engelbert
humperdinck


# Part 3: The spork

In [8]:
def boop_to_string(B: tuple[int, int]):
    """
    Converts a boop to a string

    :param tuple[int, int] B: A boop
    :returns: a/b, the fraction representing the boop
    :rtype: tuple[int, int]
    """
    return f"{B[0]}/{B[1]}"


# Tests
# print(boop_to_string((4, 7)))  # Output: '4/7'

In [9]:
def frond(n: int):
    """
    Get's the nth frond in the spork

    :param int n: Row from which frond is taken
    :returns: nth frond
    :rtype: list[tuple[int, int]]
    """
    all_kebabs = []

    def gen_kebabs(n, k=""):
        if n - 1 >= 0:
            gen_kebabs(n-1, k + "W")
            gen_kebabs(n-1, k + "E")
        else:
            all_kebabs.append(k)

    gen_kebabs(n)

    boops = [kebab_to_boop(k) for k in all_kebabs]

    return boops


# Tests
# print(frond(3))  # Output: [(1, 4), (2, 5), (3, 5), (3, 4), (4, 3), (5, 3), (5, 2), (4, 1)]

In [10]:
import numpy as np
from numpy.dtypes import StringDType


def print_spork(n: int=5):
    """
    Print spork to nth frond.

    :param int n: Depth of spork. A value of n will produce n+1 fronds.
    :returns: n-spork as array of fronds
    """

    n_spork = [frond(n) for n in range(n+1)]

    # -------------- Setup --------------
    # Find the maximum width in characters of the boops on each frond
    max_boop_widths = [max([len(boop_to_string(boop)) for boop in frond]) for frond in n_spork]
    leaf_width = max_boop_widths[-1]  # Width of leaves (boops on frond n)
    
    # This array will hold the width in cells of the sub sporks on frond i (0 <= i <= n).
    L_arr = [leaf_width]

    # Compute sub-spork widths
    while len(L_arr) < n+1:         # Adds max boop width of one row above
        L_arr.append(2 * L_arr[-1] + max_boop_widths[n-len(L_arr)])
    L_arr.reverse()
    
    # L0 is the width of the entire spork
    L0 = L_arr[0]

    # Create grid
    grid = np.empty((n+1, L0), dtype=StringDType)
    grid.fill(" ")

    # We can then recursively set each sub spork

    def set_sub_spork(c_index, i, K=""):
        # Compute a padded version of the boop
        ith_boop_width = max_boop_widths[i]
        boop_string = boop_to_string(kebab_to_boop(K))
        padded_boop_arr = list(f"{boop_string:^{ith_boop_width}}")

        # Find index of left and right of this boop
        left = c_index - ith_boop_width//2 + (1 if ith_boop_width % 2 == 0 else 0)
        right_ex =  (c_index + ith_boop_width//2) + 1 ## EXCLUSIVE


        # Apply padded boop to matrix
        grid[i, left:right_ex] = padded_boop_arr
        
        
        # Stop if we are at the bottom of the pronk
        if i == n:
            return

        # Manually specify gaps when we get towards the leaves of diagram
        if i+1 == n:
            # Second to last row
            Li2 = 0
        else:
            Li2 = L_arr[i+2]
        
        # Find half the boop width of the boops on row below
        i1th_boop_width = max_boop_widths[i+1]
        i1_half_boop = i1th_boop_width//2 

        # Compute the root location of the left and right sub-sporks
        left_centre = left - Li2 - i1_half_boop - 1
        right_centre = right_ex + Li2 + i1_half_boop - (1 if i1th_boop_width % 2 == 0 else 0)

        set_sub_spork(left_centre, i+1, K+"W")
        set_sub_spork(right_centre, i+1, K+"E")

    c_initial = L0 // 2
    set_sub_spork(c_initial, 0) 
 
    output = "".join([f'\t{"".join(map(lambda c: str(c), frond))}\t\n' for frond in grid])

    print(output)
    
    return n_spork


# This algorithm theoretically works for any size n
# Tests
print_spork(3);


	                     1/1                     	
	         1/2                     2/1         	
	   1/3         2/3         3/2         3/1   	
	1/4   2/5   3/5   3/4   4/3   5/3   5/2   4/1	



### Explanation of `print_spork`

This is actually a surprisingly tricky problem, not because it is theoretically taxing but because there are a few fiddly implementation details that must be figured out in order for the algorithm to scale to any $n$.

I opted to treat this output as a matrix of characters, and then applied a recursive algorithm to it. This way I could have extremely versatile access to my final output, and ensure every character is positioned in exactly the right place. After the algorithm is complete I simply convert the matrix into a single print output. 

#### Computation of sub-spork widths

A frond is to take up $1$ row, so the height of my matrix is $n+1$. The width is a little different. In order to get the width in characters of the entire spork, we need to know the width of the sub-sporks either side of the root fraction ($1/1$), plus the width of the root node in characters. The two sub-sporks have the same width, so we have the recurrence relation:
$$
L_{i} = 2L_{i-1} + w_{n-i}
$$
where $w_{i}$ is the maximum width in characters of the boops on frond $i$. Note that it's a little confusing here because we denote the $i$-th frond as starting at the top and moving down, i.e., $w_{n}$ is the maximum width of the leaves on this tree. The recurrence relation must be defined the other way round however for simple implementation (that is, higher $i$ means further up the tree, vertically), see the `L_arr.reverse()` to accommodate for this.

#### Recursive depth-first traversal

The method `set_sub_spork(n)` traverses the tree depth first. The function itself is responsible for setting the root node of the sub-spork it is assigned, and then calling `set_sub_spork(n)` on its left and right branches (if it has them). 

I found that the easiest way to retrieve the relevant boop at each sub-spork was to generate the kebab as we traversed the tree and then used the functions already implemented in part a to compute the boop. Traversing the `n_spork` array as we arrive at sub-sporks, or computing the boops during the traversal would be more efficient here, but this works just fine (I'm more focused on getting good formatting than performance).

It took a considerable amount of time to find the specific index adjustments required to draw a character-perfect diagram. Particularly, I had to figure out the slightly different behaviour when we have boops of even or odd width. The nested calls of `set_sub_spork(n)` expect the precise centre of that sub-spork, or a knock-on effect will eventually cause the boops on the $n$-th frond to overlap. This implementation expects the "centre" of even width boops to be the left option. Via inspection of simple examples, I determined that computing the centre of the left sub-spork always requires a subtraction of $1$, and on the right, getting the centre of an even width boop requires an adjustment of $-1$, but not for an odd-width boop.

The solution I have delivered should generalise to any $n$, and does so according to my tests. There is little practicality in increasing $n$ further than $17$ on my computer, probably because the amount of memory required increases very quickly (but this implementation can be easily modified to print line by line, bypassing this limitation). These diagrams are far too large to have any use anyway.



# Part 4: Going further

In [32]:
# To access mathematical constants
import math


def float_to_frood(q: float):
    """
    Find a pronk given a float q.

    :param float q: Floating point number used to find frood
    :returns: A frood representing q
    :rtype: list[int]
    """

    # We know we can convert boops into kebabs, then into pronks.
    # We can convert out float into a fraction.

    split_number = str(q).split(".")
    before_decimal = 0 if len(split_number) == 1 else len(split_number[1])
    
    boop = (int(q * 10**before_decimal), 10**before_decimal)
    print(boop)
    # Now we get the kebab and pronk
    kebab = boop_to_kebab(boop)
    frood = kebab_to_pronk(kebab)
    return frood

# Tests
# print(float_to_frood(54353.125))

In [12]:
def frood_to_hoopy_frood(P: list[int]):
    """
    Given a frood, finds a hoopy frood if one exists. Iterates through
    the frood and returns the array up to but excluding the first integer
    greater than 100. If there is no integer greater than 100 in the frood,
     the entire frood is returned.

    :param list[int] P: A frood
    :returns: A tuple containing subset of P at 0 and a boolean denoting 
    if the subset is indeed a hoopy frood at 1.
    :rtype: tuple[list[int], boolean]
    """

    h_frood = []

    i = 0
    found_100 = False

    while found_100 == False and i < len(P):
        s = P[i]
        if s <= 100:
            h_frood.append(s)
        else:
            found_100 = True
        i += 1

    return (h_frood, found_100)


# Tests
#print(frood_to_hoopy_frood([0, 2, 4, 5, 2, 3, 4, 5, 2, 100, 4, 5, 6, 101]))
# Output: ([0, 2, 4, 5, 2, 3, 4, 5, 2, 100, 4, 5, 6], True)
# print(frood_to_hoopy_frood([0, 2, 4, 5, 2, 3, 4, 5, 2, 100, 4, 5, 6, 10]))
# ([0, 2, 4, 5, 2, 3, 4, 5, 2, 100, 4, 5, 6, 10], False)

In [13]:
def float_to_boop(q: float):
    """
    Converts a floating point number to a boop (fraction)
    
    :param float q: floating point number to be approximated
    :returns: Boop
    :rtype: tuple[int, int]
    """

    print(f"------- Approximating {q} -------")
    frood = float_to_frood(q)
    hoopy_frood_result = frood_to_hoopy_frood(frood)
    # print(hoopy_frood_result)
    print(f"Is hoopy frood approximation: {'yes' if hoopy_frood_result[1] else 'no'}")
    
    # print(hoopy_frood_result[0])
    kebab = pronk_to_kebab(hoopy_frood_result[0])
    # print(kebab)
    boop = kebab_to_boop(kebab)
    return boop

# Tests
# print(float_to_boop(3.14))
# Output: (157, 50)
# print(float_to_boop(10000.123))
# Output (1, 1)

##### Note on large floats
The function `float_to_boop` is providing a result of `(1, 1)` when the given float has a large integer part. This is not actually a but and a result of the definition requested in the brief. When the integer part of the float is sufficiently large, finding the kebab results in repeated application of "Eep" over $100$ times, causing the approximation to become a hoopy frood, and this happens on the first number in the frood. As the definition states to truncate "immediately before" a number greater than zero, which in this case causes the hoopy frood to come back as an empty array. 

Obviously, $1$ is not a good approximation to these larger numbers. I suggest that we can fix this by simply altering the definition of a hoopy frood to include the number which stops the search. This will mean that we can get much better approximations in this scenario.

#### a)

In [14]:
print(f"Frood: {float_to_frood(math.pi)}")
print(float_to_boop(math.pi))

Frood: [3, 7, 15, 1, 292, 1, 1, 1, 2, 1, 3, 1, 14, 4, 2, 3, 1, 12, 5, 1, 5, 20, 1, 11, 1, 1, 1, 1]
------- Approximating 3.141592653589793 -------
Is hoopy frood approximation: yes
(688, 219)


With the available precision of $\pi$, the algorithm is able to find a hoopy frood approximation. This approximation is correct to 5 significant figures.
$$
\frac{688}{219} = 3.14155251142 \approx 3.141592653589793 \approx \pi
$$

#### b)

In [15]:
print(f"Frood: {float_to_frood(math.sqrt(2))}")
print(float_to_boop(math.sqrt(2)))

print("\n")

print(f"Frood: {float_to_frood(math.sqrt(3))}")
print(float_to_boop(math.sqrt(3)))

print("\n")

golden_ratio = (1 + math.sqrt(5)) / 2
print(f"Frood: {float_to_frood(golden_ratio)}")
print(float_to_boop(golden_ratio))

Frood: [1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 3, 5, 9, 4, 4, 2, 2, 3, 5, 6, 1, 1, 4]
------- Approximating 1.4142135623730951 -------
Is hoopy frood approximation: no
(1767766952966369, 1250000000000000)


Frood: [1, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 1, 1, 40, 1, 2, 2, 1, 13, 1, 3, 2, 1, 1, 5, 1, 4, 3, 1]
------- Approximating 1.7320508075688772 -------
Is hoopy frood approximation: no
(4330127018922193, 2500000000000000)


Frood: [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 37, 6, 2, 3, 8, 19, 1, 2]
------- Approximating 1.618033988749895 -------
Is hoopy frood approximation: no
(323606797749979, 200000000000000)


The algorithm was unable to calculate a hoopy frood approximation for $\sqrt{2}$, $\sqrt{3}$ and $\varphi$ with the given precision. All of these fractions are accurate up to at least $12$ significant figures.

These three froods look suspiciously like they have a repeating pattern. To check whether the level of precision on my machine is the reason the pattern is not continuing, I plugged the numbers into https://r-knott.surrey.ac.uk/fibonacci/cfCALC.html, which appears to have a smarter procedure than what I'm doing. It came back with the results:
$$
\sqrt{2} = [1; \overline{2}] \\
\sqrt{3} = [1; \overline{1, 2}] \\
\varphi = [1; \overline{1}]
$$

which agrees perfectly with my approximations. This is notation for a "continued fraction", which is what I've been calling a frood in this assignment. The bar indicates an infinite repetition of the digits below, these are infinite length froods. It's clear then that $\sqrt{2}$, $\sqrt{3}$ and $\varphi$ will never produce a hoopy frood (at least not theoretically), as the values in their frood will never get to $100$. 


I've noticed that $\pi$ is a *transcendental* number, and the three examples above are all *algebraic* irrational numbers: they can be written as the root of a polynomial with rational coefficients. Let's see if this pattern continues by checking if we get a hoopy frood approximation to another transcendental number, $e$.

In [16]:
print(f"Frood: {float_to_frood(math.e)}")
print(float_to_boop(math.e))

Frood: [2, 1, 2, 1, 1, 4, 1, 1, 6, 1, 1, 8, 1, 1, 10, 1, 1, 12, 1, 1, 10, 2, 1, 3, 1, 2, 2, 1, 2, 1, 1, 2, 1, 1, 7, 4, 1, 7]
------- Approximating 2.718281828459045 -------
Is hoopy frood approximation: no
(543656365691809, 200000000000000)


Hmm, looks like this isn't the case. Maybe there is something special about $\pi$ which means we can get such a good approximation, or maybe it's just a coincidence.

...

It turns out that *quadratic irrationals* have bounded continued fraction terms, meaning we will never find a hoopy frood for them. $\sqrt{2}$, $\sqrt{3}$ and $\varphi$ are roots for $x^2 - 2=0$, $x^2 - 3=0$ and $x^2 - x - 1=0$ respectively, so they are quadratic irrationals. A number which has a bound to it's frood is called **badly approximable**. 

$\pi$ just so happens to have a very large term in it's continued fraction very early, leading to the famous result
$$
\pi \approx \frac{355}{113}
$$

Why do we not have this result? The reason is because of our definition of a kebab. When converting the continued fraction $[3; 7, 15, 1]$ (our hoopy frood) into a kebab, we get "EEEWWWWWWWEEEEEEEEEEEEEEEW". Humans call this a **Stern–Brocot path**, and we have a slightly different definition. Our version drops trailing $1$'s from the string. So we'd have "EEEWWWWWWWEEEEEEEEEEEEEEE" representing this continued fraction. Indeed:

In [17]:
print(kebab_to_boop("EEEWWWWWWWEEEEEEEEEEEEEEE"))
# Output: (355, 113)

(355, 113)


### c)



##### 1. A friendly wumpus is mordant

Proof.

This is true. The solution is very similar to proving convergent sequences are bounded, which we proved in MAS107 last year. This is done by splitting the sequence into two parts, the part before the repeating pattern and the infinite part which is cyclical. We note that the finite part and the infinite part are bounded, so the entire sequence is bounded.

Let our wumpus be denoted $(a_n)$. Say that for $n < N$, $(a_n)$ does not exhibit a cyclical nature and $(a_n)$ for $n \geq N$ is cyclical. It is quite clear that this first part of our set is bounded. Let $A_{1, N-1} = \{a_n | 1 \leq n < N\}$ be the finite part of our sequence, then
$$
M_{1} = \text{max}(A_{N-1}) = \text{max}(a_1, a_2, \dots, a_{N-1})
$$
which is finite as all $a_n \in A_{N-1}$ are finite positive integers.

We then consider the infinite part of the sequence. We can write a single cycle in the sequence as as $A_{N, N+k} = \{a_n | N \leq n < N+k\}$, where $k$ is the length of the cycle. Note that the cyclical nature of this sequence means $A_{N, N+k} = A_{N+k, N+2k}$, in fact $A_{N, N+k} = A_{N+(q-1)k, N+qk}$ for all $q \in \mathbb{N}$, (this is the cyclical property). Put simply, we are saying that the infinite part of our sequence only contains finite distinct integers, and that these integers are exactly the set $A_{N, N+k}$. So
$$
M_{2} = \text{max}(A_{N, N+k}) = \text{max}(a_N, a_{N+1}, \dots, a_{N+k-1})
$$
Hence $\text{max}((a_n)) = \text{max}(M_{1}, M_{2})$, where $M_{1}$ and $M_{2}$ both finite positive integers and thus the wumpus is bounded from above. A friendly wumpus is mordant.
$\square$

#### 2. A mordant wumpus is friendly

We are asking if an infinite pronk being bounded implies that the pronk settles to a repeating pattern. This is the case for the examples so far, but I believe we can consider a counter-example here. 

The results we've found before follow this rule, but these are wumpuses (wumpi?) derived from specific numbers (quadratic irrational numbers specifically). There's no criteria that a wumpus be generated *from* an irrational number, just that it is a pronk of infinite length. Therefore we can posit that a non-repeating but bounded continued fraction can be generated. Here are some ways they could be made:

1. We could define a process where a new number is selected for the pronk, making sure it doesn't create a cycle.
2. We could use the digits of a transcendental number like $\pi$ to define out pronk. It is an irrational number, so it's digits never repeat, however the digits are obviously bounded by $0$ and $9$ (we can just add $1$ to each digit so there are no $0$'s as required).
3. There is something called a Thue-Morse sequence. This is a sequence of $0$'s and $1$'s, which is constructed by starting with a $0$, flipping the bits of what came before and appending it to the sequence. By repeating this process and infinite number of times, you can get a bounded infinite sequence of integers which is proven to be non-cyclical (again, adding $1$ to every digit will make this a valid wumpus).

Therefore I believe this claim to be false.

#### 3. A rampant wumpus has an infinite number of hoopy froods.

True

Proof.

If a wumpus is rampant then it is unbounded so must diverge (to $+\infty$): it has no upper bound. This means that given any finite $\alpha$, we can find an $n$ such that $\alpha < a_n$. By letting $\alpha = 100$, we discover some $N_{0} \in \mathbb{N}$ such that $100 < a_{N_0}$, and thus $\{a_n | 1 \leq n \leq N_{0}\}$ is a hoopy frood. 

$a_{N_0}$ is not an upper bound for the sequence, so there must be some $N_{1} \in \mathbb{N}$ such that $a_{N_0} < a_{N_1}$, which implies that $\{a_n | 1 \leq n < N_1\}$ is another, different hoopy frood. This process can be repeated an infinite number of times, so we can create an infinite number of hoopy froods if our wumpus is rampant. $\square$

#### 4. Every wumpus is mordant

Similarly to claim 2, a wumpus is not descried as necessarily being a representation of a specific irrational number (just that an irrational number can be represented as a wumpus, and a wumpus is simply a pronk of infinite length).

Obviously then this is false, just pick any unbounded increasing integer sequence and we have ourselves a rampant wumpus. For example $\{a | a \in \mathbb{Z}^+\}$ can be a wumpus, which is rampant because it has no upper bound.

# Acknowledgements

https://r-knott.surrey.ac.uk/fibonacci/cfCALC.html - used to find the exact continued fractions for the example.

https://en.wikipedia.org/wiki/Diophantine_approximation

https://en.wikipedia.org/wiki/Irrationality_measure

https://www.geeksforgeeks.org/python/string-alignment-in-python-f-string/

https://math.stackexchange.com/questions/4112417/continued-fraction-representation-of-quadratic-irrationals

https://stackoverflow.com/questions/9476797/how-do-i-create-character-arrays-in-numpy

https://stackoverflow.com/questions/3898572/what-are-the-most-common-python-docstring-formats

https://numpy.org/devdocs/user/basics.strings.html

https://en.wikipedia.org/wiki/Engelbert_Humperdinck_(singer)

https://editor.codecogs.com/

## Acknowledgement of AI Assistance


### 1. Declaration
Please mark the relevant box with an 'x':
- [ ] I did **not** use any AI assistance for this assignment.
- [x] I **did** use AI assistance, as detailed below.

### 2. AI Services Used
ChatGPT, Gemini Chat (3.1 Pro), Gemini AI overview (automatic when google searching)

### 3. Description of Usage

In general, I used the automatic AI response from a number of google searches.

Part 3.a: asked Gemini to create a program for outputting the spork. I didn't use this however as 
I decided it would be easier to create it myself than try to debug what it had made (it didn't work).

Part 4.b: I asked ChatGPT to identify what area of maths this was, which allowed me to do further
research on the topic. I asked ChatGPT if the hoopy frood for pi was a coincidence or not. I showed
ChatGPT my code for converting a kebab to a boop and asked if there were any problems with it - 
trying to ascertain if my estimation of 688/219 for pi was an error.

Part 4.c.1: I asked ChatGPT if my first suggestion to construct a non-cyclical wumpus bounded would 
work - unfortunately it wouldn't give me a convincing answer. I asked Gemini the same thing and was happy 
with the reasoning it provided.

### 4. Statement of Understanding
By submitting this assignment, I confirm that:
1.  I have tested and verified that all code behaves as expected.
2.  I fully understand the logic and syntax of the code submitted.
3.  I am capable of explaining the function of every line if asked.
4.  The written text of this report is my own work.

# Sandbox

In [18]:
# For other code used when writing your notebook.
# This part will not be marked.
# This part can be blank in the final submission.